### PR TITLE
Implement imposition of hanging node constraints and periodic boundaries for CG discretizations

### LINKED ISSUE #385: Correctly impose hanging node and periodic boundary constraints for Poisson / Structure modules with CG discretization
It seems as if hanging node and periodic boundary constraints are not taken into account when setting up the `AffineConstraints` object in the `Poisson` module, see https://github.com/exadg/exadg/blob/811895767191b4170beed035140641b024f7424c/include/exadg/poisson/spatial_discretization/operator.cpp#L135-L165

and in the `Structure` module, see https://github.com/exadg/exadg/blob/41f8752a19a667b64009fcd6f4325a087afe8baa/include/exadg/structure/spatial_discretization/operator.cpp#L129-L157

### PR DESCRIPTION
**Pull request by nfehn**
closes #385

Note: The present PR can be considered a "bugfix", because we didn't assert before that hanging nodes and periodic boundaries were actually not working correctly with continuous Galerkin discretizations. 

The two utility functions could be relevant for the deal.II library. Does deal.II already provide such functionality @peterrum @kronbichler? I could not find it.

thanks to @bergbauer for your help how to construct a `DoF::cell_iterator` from a `Tria::cell_iterator`. The deal.II docu (https://www.dealii.org/developer/doxygen/deal.II/classDoFCellAccessor.html#a3fdfeda1dae3c8c79aab8f8693d773f4) is quite hard to understand in my opinion.

To test the periodic boundary condition implementation, I circumvented the `GridUtilities::create_triangulation()` functionality for global-coarsening (see problem described in issue #387), and used local-smoothing multigrid instead, by constructing the fine triangulation and related periodic face pairs explicitly in application.h (for the poisson/sine app). The numerical solution was wrong with the old code version, i.e. it was not periodic, and also iteration counts reduced drastically (from 33.2 to 5.9 for the example I have chosen) with the new implementation.

Note that this PR only addresses the setup of the AffineConstraints objects. Not all operators will work correctly with the changes of this PR (because of `read_dof_values_plain()` used at some places in the code, see issue #383).

### CODE REVIEW

**Review by peterrum — APPROVED** [review]
I like this solution with `transform_periodic_face_pairs_to_dof_cell_iterator()`. I normally run the function that collects periodic faces twice (once for Tria and once for DoFHandler).

File: include/exadg/grid/grid_utilities.h [hunk 1]
```diff
@@ -90,6 +90,86 @@ get_mesh_smoothing(bool const use_local_smoothing_multigrid, ElementType const &
...
+    PeriodicFacesDoF;
+
+  PeriodicFacesDoF periodic_faces_dof;
+
+  for(auto it : periodic_faces)
+  {
+    dealii::GridTools::PeriodicFacePair<typename dealii::DoFHandler<dim>::cell_iterator>
+      face_pair_dof_hander;
+
+    face_pair_dof_hander.cell[0] = typename dealii::DoFHandler<dim>::cell_iterator(
```
[the remark this answers is not in the document]
Comment (nfehn) [reply include/exadg/grid/grid_utilities.h:117]: Do you mean I should try to use the deal.II function here?

[the remark this answers is not in the document]
Comment (kronbichler) [reply include/exadg/grid/grid_utilities.h:117]: This is a recent function in deal.II, so we have probably many places in ExaDG that could be converted to this format.

File: include/exadg/grid/grid_utilities.h [hunk 2]
```diff
@@ -90,6 +90,86 @@ get_mesh_smoothing(bool const use_local_smoothing_multigrid, ElementType const &
...
+}
+
+/**
+ * This function can be seen as some form of "copy constructor" for periodic face pairs,
+ * transforming the template argument of dealii::GridTools::PeriodicFacePair from
+ * DoFHandler::cell_iterator to Triangulation::cell_iterator.
+ */
+template<int dim>
+std::vector<dealii::GridTools::PeriodicFacePair<typename dealii::Triangulation<dim>::cell_iterator>>
+transform_periodic_face_pairs_to_tria_cell_iterator(
```
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/exadg/grid/grid_utilities.h:141]: Do we need the function, or could one even make the assignment to get an object of the type
```c++
  std::vector<dealii::GridTools::PeriodicFacePair<typename dealii::Triangulation<dim>::cell_iterator>>
```
by assigning it an object of the type
```c++
  std::vector<dealii::GridTools::PeriodicFacePair<typename dealii::DoFhandler<dim>::cell_iterator>>
```
This would be an upcast, so I would have assumed that the `std::vector` propagates through the automatic assignment operators.

File: include/exadg/poisson/spatial_discretization/operator.cpp [hunk 3]
```diff
@@ -136,6 +138,20 @@ Operator<dim, n_components, Number>::distribute_dofs()
   {
     affine_constraints.clear();
```
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/exadg/poisson/spatial_discretization/operator.cpp:140]: Yes, we need `AffineConstraints::reinit(const IndexSet &locally_relevant_constraints)`, because otherwise `AffineConstraints` will allocate a lookup table of size proportional to the global size of the problem accumulated over all MPI processes, which will make simulations with >100m unknowns (or something around that size) to run out of memory on any number of parallel processes. The initialization to the `locally_relevant_dofs` makes sure that the lookup table only uses this selected number of DoFs, i.e., memory consumption on each MPI process is proportional to the number of unknowns on each process.

**Review by kronbichler — APPROVED** [review]
OK apart from my comments.

File: include/exadg/grid/grid_utilities.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (peterrum) [reply include/exadg/grid/grid_utilities.h:141]: How would that work? The `DoFCellAccessor` needs a reference to the underlying `DoFHandler`!?

[the remark this answers is not in the document]
Comment (peterrum) [reply include/exadg/grid/grid_utilities.h:141]: > yes, not used currently, but I did it because it was very easy to obtain after I implemented the other function

I would remove it. I cannot think of an instance where one would need that kind of translation.

[the remark this answers is not in the document]
Comment (kronbichler) [reply include/exadg/grid/grid_utilities.h:141]: I mean just the conversion `DoFHandler::cell_accessor` -> `Triangulation::cell_accessor` (currently not used), not the other way around.

[the remark this answers is not in the document]
Comment (nfehn) [reply include/exadg/grid/grid_utilities.h:141]: I will test the suggestion made here https://github.com/exadg/exadg/pull/388/files#r1137469215 (just out of curiosity), and remove this utility function, since you do not seem to see a value in that function if I got it right.

[the remark this answers is not in the document]
Comment (nfehn) [reply include/exadg/grid/grid_utilities.h:141]: > This would be an upcast, so I would have assumed that the std::vector propagates through the automatic assignment operators.

This does not seem to work!? Even if I do the assignment element by element (by a range based for-loop), the automatic assignment does not seem to work. (Note that there is a typo in the suggestion above: DoFhandler -> DoFHandler).

Should I remove the second utility function @kronbichler?

File: include/exadg/grid/grid_utilities.h (hunk 1 above)
[the remark this answers is not in the document]
Comment (nfehn) [reply include/exadg/grid/grid_utilities.h:117]: Using
```c++
face_pair_dof_hander.cell[0] = it.cell[0].as_dof_handler_iterator(dof_handler);
```
I get the error message "‘class dealii::TriaIterator<dealii::CellAccessor<2, 2> >’ has no member named ‘as_dof_handler_iterator’". 

Is "Active" the problem here or do I have to call the function differently?

File: include/exadg/poisson/spatial_discretization/operator.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (nfehn) [reply include/exadg/poisson/spatial_discretization/operator.cpp:140]: @kronbichler I added these lines. However, I am not sure whether this is needed if we have periodic boundaries but no hanging nodes (I assumed so)? Another question is whether we should do it always, or only if `hanging_nodes or periodic_faces`?

**Comment by nfehn** [comment]
I added the same functionality for the `Structure` module. Note that we have a mass operator there with a separate `AffineConstraints` object (background: for unsteady problems, we need the inverse mass operator in the time integrator, where we solve a linear system of equations for the mass matrix system (Continuous Galerkin discretization)). The time derivative operator (mass matrix) in the nonlinear residual is evaluated within the `NonLinearOperator` (using the `AffineConstraints` object with Dirichlet boundary conditions).

**Comment by nfehn** [comment]
The utility function introduced in this PR is required for PR #389, so it would be great to make progress here.

**Comment by nfehn** [comment]
@kronbichler your input would be required here.

File: include/exadg/grid/grid_utilities.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/exadg/grid/grid_utilities.h:138]: Yes, I would remove this function. And for the record, I tested the automatic conversion as `vector_of_tria_iterators(vector_of_dofh_iterators.begin(), vector_of_dofh_iterators.end())` and realize that the `PeriodicFacePair` would need a custom constructor because the compiler would not automatically convert the iterators. So it would be more work than I suggested.

File: include/exadg/poisson/spatial_discretization/operator.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (kronbichler) [reply include/exadg/poisson/spatial_discretization/operator.cpp:140]: Yes, always do it as soon as `AffineConstraints` gets to hold any constraint. It can be skipped if you do not insert any DoF, though.

**Review by kronbichler — APPROVED** [review]
This looks good, thank you. See two comments below.
